### PR TITLE
Windows Code::Blocks project remake

### PR DESCRIPTION
**Pull request by Forsari0**
- Now all dependencies located at `WinDepend/` folder, no more mess like in Release build dependencies are in `WinDepend/` dir and for Debug build they are in `3rdparty/` folder
- Now all headers should reside at `WinDepend/include` dir and all libraries should be at `WinDepend/lib` folder, no more mess like `WinDepend\SDLImg_binVC\SDL_image-1.2.12\lib\x86`, `WinDepend\SDLttf\SDL_ttf-2.0.11\lib\x86`, `WinDepend\SDL-1.2.15\lib`, etc.
- Now each build target has separate `object build` folder
- All targets, which depend on some lib\libs, now use shared (`dll` in windows world) libraries for linking
- Add `-D_WINDOWS -D_MINGW -D_WIN32 -DWIN32 -D__MINGW__` flags
- Add `CB_BUILD` define for using in `src/version.cpp` (see pre-build step below)
- Add more targets for `Debug` and `Release` builds:
  - 1) `Lua`
  - 2) `SDL, Sound`
  - 3) `SDL, Localized`
  - 4) `SDL, Sound, Localized, Lua`
- Add pre-build step for all targets (like in `.vcxproj` in `msvc140/`) - get most recent commit tag and print it to the `src/version.h` header file
  
  > Also, there's a error in `msvc140/getversion.cmd`, at second square brackets - regex `[0-9]*.[0-9]*` should be `[0-9]*.[0-9A-Z]*` for getting correct result - **0.B**
- Add pre-build step for all `Lua` targets - generate `catabindings.cpp` from `generate_bindings.lua` file

### CODE REVIEW

**Comment by narc0tiq** [comment]
>> Also, there's a error in `msvc140/getversion.cmd` [...]

Which still seems wrong to me -- in particular, your variant doesn't match "0.B" at all (though it does match "**B.0**"). The Makefile uses `[0-9A-Z]*.[0-9A-Z]*`, which matches things like "0.B", "B.0", and even "ABC+XYZ" -- it's exceptionally greedy, but limited by the fact we don't use tags for anything other than releases.

The simplest variant, to the best of my knowledge, should be `[0-9]+\.[0-9A-Z]` -- all our versions so far have matched this format and I would assume they will continue to do for the foreseeable future.

**Comment by Forsari0** [comment]
Yep, my bad, fix first post.
Also what is particular reason for using `--match` flag? In our case, we just need to get **last** commit tag and `git describe` makes it very well even without `--match` flag at all.
I just don't use this flag in C::B projects pre-build step and it works like a charm.

**Comment by narc0tiq** [comment]
The reason for `--match` is because we may want to use the tags for other things, too -- although, while we have done so experimenally a couple of times, it's fallen out of favor. It's probably pretty safe to do without.

File: CataclysmWin.cbp [hunk 1]
```diff
@@ -7,170 +7,309 @@
 		<Option compiler="gcc" />
 		<Build>
 			<Target title="Release">
-				<Option output="Cataclysm" prefix_auto="1" extension_auto="1" />
+				<Option output="CataclysmWin" prefix_auto="1" extension_auto="1" />
```
Comment (KA101) [inline CataclysmWin.cbp]: Marginally irked at being tagged as "CataclysmWin", seems unnecessary to mark as Not Linux.  :-/

Comment (Forsari0) [inline CataclysmWin.cbp]: @KA101, but part of C::B project file name `CataclysmWin.cbp` is `Win`, which states that this project for Windows users only, don't it?

Comment (KA101) [inline CataclysmWin.cbp]: I'll take a look: as I read that code, the executable would be "CataclysmWin" rather than Cataclysm, and that seemed off..

**Comment by KA101** [comment]
OK, this gives an error message (http://pastebin.com/vDw0d7xg) when I start a build but apart from that seems to work OK.  Since we're looking at a release shortly, gonna ask that we sort that out before merging this.

**Comment by narc0tiq** [comment]
I'm not seeing an error message. That's the command line: "run `git
--version` and, if it fails, `echo 'git' command not found`". It didn't
fail, the error wasn't printed. Presumably this is to ensure the build
doesn't fail due to an unexpected nonzero return, but that's just a guess.

**Comment by KA101** [comment]
OK then, Narc, thanks.  As it stands, the project file I had before this just **built** when I pushed he build button, rather than printing out an extra command that may or may not run.

**Comment by bkentel** [comment]
Why is the lib *path* even in the project file? Should that not be set locally on a per user basis globally in codeblocks? I know my libs aren't in WinDepend for example.

**Comment by Forsari0** [comment]
@KA101, I can't find possibility to hide output of pre-build step commands to `Build log` panel. However, it is not necessary, because **all** errors, if there's any, goes to `Build messages` panel. So, if `Build messages` shows `Build finished: 0 error(s), ...` there's no errors at all.
@bkentel, I think it's much more easy for newcomers to use *ready-to-use infrastructure*. Of course, you can set your own path to `lib` and `include` directories in global C::B compiler settings, but for those, who don't like change global settings because of just one project, these predefined paths would be useful.

**Comment by macrosblackd** [comment]
I've tested this with the latest code and even created a pre-packaged win64 download for the needed libs.

**Comment by kevingranade** [comment]
This fell through the cracks, the changes look correct, and last report is from macrosblackd that it works, can I get confirmation that it still works, or does so with minor fixes?

**Comment by illi-kun** [comment]
I've tested "Release" & "Debug" and these modes build fine but I didn't check "Lua/SDL/Sound" builds which is the most important part of this PR.

**Comment by macrosblackd** [comment]
I tested with lua, sound, tiles and everything worked

**Comment by macrosblackd** [comment]
When I get some extra time this weekend I will rebase this PR on master and handle the conflicts myself and then create a new PR

**Comment by macrosblackd** [comment]
I have created a new PR with this branch rebased on the latest master (#13324)

**Pull request merged**